File: 3sum/Program.cs

int [] nums = [-1, 0, 1, 2, -1, -4];

var res = ThreeSum(nums);
foreach(var tuple in res)
    Console.WriteLine($"({tuple.Item1}, {tuple.Item2}, {tuple.Item3})");

Console.WriteLine(string.Join(", "), res);

(int, int, int)[] ThreeSum(int[] nums) 
{
    if (nums.Length < 3) return [];
    var result = new HashSet<(int,int,int)>();

    // Existing numbers between nums[i] and nums[j]
    for(int i=0; i<nums.Length-2; i++)
    {
        int target = 0 - nums[i];
        var hashSet = new HashSet<int>();
        
        for(int j=i+1; j<nums.Length; j++)
        {
            if(hashSet.Contains(target - nums[j]))
            {
                int[] arr = [nums[i], target - nums[j], nums[j]];
                Array.Sort(arr);
                result.Add((arr[0], arr[1], arr[2]));
            }
            else
            {
                hashSet.Add(nums[j]);
            }
        }
    }
    return result.ToArray();
}

File: 3sum/program.py
class Solution(object):
  def threeSum(self, nums):
    nums.sort()
    LEN = len(nums)
    res = []
    for a in range(LEN):
        if a > 0 and nums[a] == nums[a-1]:
            continue
        
        l, r = a + 1, LEN - 1
        while l < r:
            sum = nums[a] + nums[l] + nums[r]
            if sum < 0:
                l += 1
            elif sum > 0:
                r -= 1
            elif sum == 0:
                res.append([nums[a], nums[l], nums[r]])
                l += 1
                while l < r and nums[l] == nums[l-1]:
                    l += 1
    return res

# Test Program
nums = [1, -2, -2, 1, 0,  0, 5, 1]
print(Solution().threeSum(nums))
# [[-2, 1, 1]]
